Adds kubectl-ccsecret plugin file

rewanth1997 opened the pull request:
Requesting to add `ccsecret` plugin to krew index.

I hope I followed all guidelines, and I did verify installing it locally.

Link to source code: https://github.com/rewanth1997/kubectl-ccsecret

```
wget https://raw.githubusercontent.com/rewanth1997/kubectl-ccsecret/master/deploy/krew/plugin.yaml
kubectl krew install --manifest=./plugin.yaml
```

Repo link: https://github.com/rewanth1997/kubectl-ccsecret

Please review the code. I'm happy to make the necessary changes if any.

Regards,
Rewanth

rewanth1997 commented:
/assign @corneliusweig @ahmetb

rewanth1997 commented:
@corneliusweig @ahmetb Feel free to suggest a better name to improve the discoverability of this plugin.

rewanth1997 commented:
@luksa Similar to last time (https://github.com/kubernetes-sigs/krew-index/pull/254), I developed one more plugin. Please review it and provide your valuable inputs.

corneliusweig commented on plugins/ccsecret.yaml:
```diff
@@ -0,0 +1,62 @@
+        arch: amd64
+  version: v1.0.0
+  homepage: https://github.com/rewanth1997/kubectl-ccsecret
+  caveats: |
```
Please reserve `caveats` for things like dependencies or prerequisites.

rewanth1997 replied:
Noted and fixed

corneliusweig commented on plugins/ccsecret.yaml:
```diff
@@ -0,0 +1,62 @@
+  homepage: https://github.com/rewanth1997/kubectl-ccsecret
+  caveats: |
+    Usage:
+      kubectl ccsecret <COMMAND> <FLAGS>
```
This is shown by default by `krew info`

rewanth1997 replied:
Done

corneliusweig commented on plugins/ccsecret.yaml:
```diff
@@ -0,0 +1,62 @@
+      kubectl ccsecret <COMMAND> <FLAGS>
+
+    Documentation:
+      https://github.com/rewanth1997/kubectl-ccsecret
```
There's already the `homepage` field.

rewanth1997 replied:
Removed it.

corneliusweig commented on plugins/ccsecret.yaml:
```diff
@@ -0,0 +1,62 @@
+    Enter value for key1: ******
+    Enter value for key2: ******
+
+    Create docker-registry secret with --docker-password:
+    $ kubectl ccsecret docker-password my-secret --docker-password
+    Enter value for docker password: ******
```
One example should be enough. I don't think it adds much to have two almost identical ones.

rewanth1997 replied:
Sure, makes sense.

corneliusweig commented on plugins/ccsecret.yaml:
```diff
@@ -0,0 +1,62 @@
+    input from the console to prevent information leakages via terminal history,
+    shoulder surfing, etc
+    
+    More on https://github.com/rewanth1997/kubectl-ccsecret
```
The `homepage` field is shown when calling `krew info`. No need to repeat yourself.

rewanth1997 replied:
Yeah, done.

corneliusweig reviewed (commented):
I think this plugin is original and useful.

I would usually argue that creating resources from CLI is bad practice, but secrets may be an exception. Let's also hear another opinion from @chriskim06.

Regarding naming: `ccsecret` is fine regarding our naming rules. However, I find it not very intuitive or memorable. WDYT about something like `blind-secret`?

rewanth1997 commented on plugins/ccsecret.yaml:
```diff
@@ -0,0 +1,62 @@
+    bin: kubectl-ccsecret
+    files:
+    - from: "*"
+      to: "."
```
Done

rewanth1997 commented:
@corneliusweig Thanks for the quick review and I'm glad you liked the idea. Most of the issues you pointed in the manifest file are related to redundancies and I fixed them all.

I spent more time brainstorming for an apt name than I spent building this plugin but couldn't come up with any handy name.

`blind-secret` is a good idea but since we are creating secrets, I think we should have `create` keyword in the plugin name. I thought of `create-hidden-secret`, `create-console-secret` but these are too lengthy to type, so I changed the name to `ccsecret`, a short form of `create console secret`. I think maybe we can make it `create-blind-secret` or `cbsecret` or `create-bsecret`. I don't know, this is very confusing.

corneliusweig commented on plugins/ccsecret.yaml:
```diff
@@ -0,0 +1,41 @@
+    input from the console to prevent information leakages via terminal history,
+    shoulder surfing, etc
+
+    Examples:
```
nit
```suggestion
    Example:
```

rewanth1997 replied:
Fixed.

corneliusweig commented on plugins/ccsecret.yaml:
```diff
@@ -0,0 +1,41 @@
+    Examples:
+
+    Create generic secret with --from-literal:
+    $ kubectl ccsecret generic my-secret --from-literal key1 --from-literal key2
+    Enter value for key1: ******
+    Enter value for key2: ******
```
```suggestion
    $ kubectl ccsecret generic my-secret --from-literal key
    Enter value for key: ******
```
I think the second item does not add much. But you can also leave it there.

rewanth1997 replied:
True. Done.

corneliusweig commented on plugins/ccsecret.yaml:
```diff
@@ -0,0 +1,41 @@
+        arch: amd64
+  version: v1.0.0
+  homepage: https://github.com/rewanth1997/kubectl-ccsecret
+  shortDescription: Create kubectl secrets by taking sensitive input from console
```
No need to mention `kubectl`, this is redundant. Maybe keep the description a bit more abstract, like so:
```suggestion
  shortDescription: Create secrets with improved privacy
```

rewanth1997 replied:
Yeah, this sounds more rhetorical. Done.

corneliusweig reviewed (commented):
The problem with `ccsecret` is that the name does not transport the idea behind the plugin. So I'd recommend to spend more time on a good plugin name. You could also be more poetic and call it `whisper-secret` or so.

Btw, I like a lot that you simply wrap `kubectl` with your plugin. I wish more plugins were designed in that way.

---

Idea for the next version: add an option that requires to type secrets twice, so that you can spot typos.

rewanth1997 commented:
@corneliusweig I agree, even I didn't like the `ccsecret` name. `whisper-secret` sounds promising and aptly conveys the idea of the plugin IMO. I really liked it 😄 Maybe we can wait for one more day to see if others have any recommendations. @ahmetb @luksa

Thanks for the compliment on the implementation, I'm glad you had a glance at the code. The modular coding style and structure make it easy to integrate other flags (if any)  as well to the list of `whisper-secret`, whenever required.

---
In the next version, we can add an option like `--check/-c` that allows the user to enter the secret value twice to check typos. This will be a good addition.

rewanth1997 commented:
Since we are considering security POV, we will set the default setting to prompt the input box twice. In the next release, we can add a flag like (`--ignore-typo/-i`) that allows the user to accept the typo risk and prompts the input field only once.

ahmetb commented:
Honestly the name is ok with me, I might've called it `mksecret` or something like that otherwise. Right now the `ccsecret` name indeed doesn’t convey the meaning of the plugin well ("create secrets with secure input prompt"), and therefore it'll be less discoverable, but it is acceptable.

rewanth1997 commented:
Thanks @ahmetb. `mksecret` is also good but it doesn't convey the meaning of the plugin well. I liked @corneliusweig recommendation `whisper-secret` which kind of conveys the meaning of the plugin but it's a bit long. Do you have any other handy names coming to your mind that are short and also might convey the meaning of the plugin? @ahmetb @corneliusweig

ahmetb commented:
I just tested this plugin and noted it doesn't work on darwin, but you provided binary release. So maybe there's need for more polishing here.

I recommended `mksecret` because it's like `make-secret` but shorter. It might be just me :) You can call it `type-secret` or `input-secret`.

ahmetb commented:
/hold cancel
/lgtm
/approve

Also, please consider setting up [Krew release automation](https://krew.sigs.k8s.io/docs/developer-guide/release/automating-updates/) which helps you skip manually making updates to your Krew manifests on each new version and send a pull request. It’s a GitHub workflow bot that can run every time you push a new tag. Those PRs are automatically approved.
